File: SWEA/D4/3124. 최소 스패닝 트리/최소 스패닝 트리.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import java.util.StringTokenizer;

public class Solution {

    static int v, e;
    static List<int[]> list;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int T = Integer.parseInt(br.readLine());
        for (int tc=1; tc<=T; tc++) {
            StringTokenizer st = new StringTokenizer(br.readLine());
            v = Integer.parseInt(st.nextToken());
            e = Integer.parseInt(st.nextToken());

            list = new ArrayList<>();
            for (int i=0; i<e; i++) {
                st = new StringTokenizer(br.readLine());
                int x = Integer.parseInt(st.nextToken());
                int y = Integer.parseInt(st.nextToken());
                int z = Integer.parseInt(st.nextToken());
                list.add(new int[] {x, y, z});
            }
            list.sort(Comparator.comparingInt(x -> x[2]));

            System.out.println("#" + tc + " " + go());
        }
    }

    static long go() {
        UnionFind uf = new UnionFind(v);
        int cnt = 0;
        int idx = 0;

        long total = 0;
        while (cnt < v-1) {
            int[] edge = list.get(idx++);
            if (uf.union(edge[0], edge[1])) {
                total += edge[2];
                cnt++;
            }
        }

        return total;
    }

    static class UnionFind {
        int[] parent;

        public UnionFind(int size) {
            this.parent = new int[size+1];
            for (int i = 0; i <= size; i++) {
                parent[i] = i;
            }
        }

        public int find(int x) {
            while (x != parent[x]) {
                parent[x] = parent[parent[x]];
                x = parent[x];
            }

            return x;
        }

        public boolean union(int u, int v) {
            int uRoot = find(u);
            int vRoot = find(v);

            if (uRoot == vRoot) {
                return false;
            } else {
                if (uRoot < vRoot) {
                    parent[vRoot] = uRoot;
                } else {
                    parent[uRoot] = vRoot;
                }
            }

            return true;
        }
    }
}

File: 백준/Gold/1197. 최소 스패닝 트리/최소 스패닝 트리.py
import sys

input = sys.stdin.readline

v, e = map(int, input().split())
graph = [tuple(map(int, input().split())) for _ in range(e)]
graph.sort(key=lambda x: x[2])

class UnionFind:
    def __init__(self, size):
        self.parent = [i for i in range(size+1)]
        
    def find(self, x):
        while x != self.parent[x]:
            self.parent[x] = self.parent[self.parent[x]]
            x = self.parent[x]
        return x
    
    def union(self, u, v):
        u_root = self.find(u)
        v_root = self.find(v)
        
        if u_root < v_root:
            self.parent[v_root] = u_root
        elif u_root > v_root:
            self.parent[u_root] = v_root
        else:
            return False

        return True
    
uf = UnionFind(v)
ans = 0
cnt = 0
for a, b, c in graph:
    if uf.union(a, b):
        ans += c
        cnt += 1
    
    if cnt == v-1:
        break

print(ans)